docs/extending/code-overview/evaluation.rst

```diff
@@ -135,30 +136,40 @@ step of this process. See also `The Standard Evaluation Sequence
 Gather information from ``Head`` and gather its leaves
 ------------------------------------------------------
 
-This is done using these attributes in ``Head``:
+What appears in this section follows largely what was in the section
+A.4.1 "The Standard Evaluation Sequence" of the Mathematica Book for
+version 5.
+
+If the expression is a literal value, e.g. ``Number``, ``String``, ``Image``, etc.) then leave it unchanged and return.
+
+Evalute ``Head`` and get its attributes.
+
+Depending the following attributes
 
 * ``HoldFirst``,
 * ``HoldAll``,
 * ``HoldRest``
+* ``HoldAllComplete``
```

Comment:
@mmatera note that while this is mentioned in A.4.1 The Standard Evaluation Sequence and it also appears in our code, it is _not_ mentioned in the comments.  So please double check this section to make sure I haven't gotten things wrong, which I occasionally do.

Overall lets _not_ revise the comments in the code until we are happy here with this description. _After_ that is done would be the time to then go back to the code and fix up the comments (and probably also the code if it is lacking, could be improved or just plain wrong.)

Reply:
I think it is exactly as this part of the evaluation works. I was not more verbose on this point in the code because I thought that the meaning of each of these attributes were described in the docstrings of the corresponding symbols.

Reply:
Ok. But in the RsT we should be very verbose. And in that process I'd like to understand where things look to be different such as that part where there is supposed to be a test for a literal, or to quote from the page 1045:

> If the expression is a raw object (e.g. `Integer`, `String`, etc.) leave it unchanged

The Mathics rewrite_apply_eval_step doesn't do or at least not before anything else is done. So the question is, should it be made to do so which would be a performance win, or would that create a bug somewhere.

What Wolfram writes goes back over a decade so things could have changed. And even when it was written, it was not presented as something you could hold it accountable as a truly accurate statement, even aside from the vagueness of whatever "raw object" means.

Reply:
Right now, if there is not an attribute that specified the opposite, each element of the expression is "evaluated". For "Raw" objects (strings, numbers, or Images) it means to call the ``evaluate`` method of these objects, which just check if the evaluation was stopped. My guess is that we should change this, in a way that only "evaluable" objects be evaluated. However, I have to say that adding an ``if`` statement in ``rewrite_apply_eval_step`` to evaluate the evaluable leaves didn't produce any observable improvement in performance.

Reply:
At some point I'll benchmark to see what's up here.